assertsprocessor/sampler.go

```diff
@@ -0,0 +1,195 @@
+	errorQueue *TraceQueue
+}
+
+func (tQ *traceQueues) errorTraceCount() int {
+	return len(tQ.errorQueue.priorityQueue)
+}
+
+func (tQ *traceQueues) slowTraceCount() int {
+	return len(tQ.slowQueue.priorityQueue)
+}
+
+type sampler struct {
+	logger               *zap.Logger
+	config               *Config
+	thresholdHelper      *thresholdHelper
+	topTracesMap         *sync.Map
+	healthySamplingState *sync.Map
+	traceFlushTicker     *clock.Ticker
+	nextConsumer         consumer.Traces
+	stop                 chan bool
+	requestRegexps       *map[string]*regexp.Regexp
+}
+
+type traceSummary struct {
+	hasError        bool
+	isSlow          bool
+	slowestRootSpan ptrace.Span
+	requestKey      RequestKey
+	latency         float64
+}
+
+func (s *sampler) sampleTrace(ctx context.Context,
+	trace ptrace.Traces, traceId string, spanSet *resourceSpanGroup) {
+	summary := s.getSummary(traceId, spanSet)
+	item := Item{
+		trace:   &trace,
+		ctx:     &ctx,
+		latency: summary.latency,
+	}
+	if summary.hasError || summary.isSlow {
+		// Get the trace queue for the entity and request
+		pq := s.getTraceQueues(summary.requestKey)
+		if summary.hasError {
+			s.logger.Info("Capturing error trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.errorQueue.push(&item)
+		} else {
+			s.logger.Info("Capturing slow trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.slowQueue.push(&item)
+		}
+	} else if len(spanSet.rootSpans) > 0 {
+		// Capture healthy samples based on configured sampling rate
+		state, _ := s.healthySamplingState.LoadOrStore(summary.requestKey.AsString(), &periodicSamplingState{
+			lastSampleTime: 0,
+			rwMutex:        &sync.RWMutex{},
+		})
+		samplingState := state.(*periodicSamplingState)
+		s.logger.Info("Found Normal trace",
+			zap.String("request", summary.requestKey.AsString()),
+			zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+			zap.Float64("latency", summary.latency),
+			zap.Time("lastSampleTime", time.Unix(samplingState.lastSampleTime, 0)))
+		if samplingState.sample(s.config.NormalSamplingFrequencyMinutes) {
+			s.logger.Info("Capturing normal trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq := s.getTraceQueues(summary.requestKey)
+
+			// Push to the latency queue to prioritize the healthy sample too
+			pq.slowQueue.push(&item)
+		}
+	}
+}
+
+func (s *sampler) getSummary(traceId string, spanSet *resourceSpanGroup) *traceSummary {
+	summary := traceSummary{}
+	summary.hasError = false
+	summary.isSlow = false
+	maxLatency := float64(0)
+	summary.hasError = summary.hasError || spanSet.hasError()
+	entityKey := buildEntityKey(s.config, spanSet.namespace, spanSet.service)
+	for _, rootSpan := range spanSet.rootSpans {
+		request := getRequest(s.requestRegexps, rootSpan)
```

Comment:
We should skip and continue the loop if `request` is an empty string

Reply:
> We should skip and continue the loop if `request` is an empty string

This is really a gap in mapping the request from the trace side. We do have the threshold at the service level which could be applied. A better option may be to default the request to the span name if the regexp expressions don't yield a value.